<launch>
    <!-- Detection configuration -->
    <arg name="weights" default="$(find qm_vision)/models/handle_detection.pt"/>
    <arg name="confidence_threshold" default="0.75"/>
    <arg name="iou_threshold" default="0.45"/>
    <arg name="maximum_detections" default="1"/>
    <arg name="threshold_l"      default="30"/>
    <arg name="threshold_u"      default="255"/>

    <!-- Visualize using OpenCV window -->
    <arg name="view_image" default="true"/>
    <!-- Publish images to ros topic -->
    <arg name="publish_image" default="true"/>
    
    <!-- Input data topics -->
    <arg name="input_image_topic" default="/camera/color/image_raw"/>
    <arg name="camera_info_topic" default="/camera/depth/camera_info"/> 
    <arg name="input_depth_topic" default="/camera/depth/image_rect_raw"/>

    <!-- Module topics-->
    <arg name="output_image_topic" default="/yolo/image"/>
    <arg name="output_mask_topic" default="/yolo/mask"/>
    <arg name="output_depth_topic" default="/yolo/depth_registered"/>
    <arg name="output_masked_depth_nr_topic" default="/yolo/depth_masked"/>
    <arg name="output_masked_depth_topic" default="/yolo/masked_depth_registered"/>  
    <arg name="output_pc_topic" default="/handle/points"/>
    <arg name="output_centroid_topic" default="/handle/centroid"/>
    <arg name="output_orientation_topic" default="/handle/orientation"/>

    <node pkg="qm_vision" name="detect" type="detect.py" output="screen">
        <param name="weights" value="$(arg weights)"/>
        <param name="confidence_threshold" value="$(arg confidence_threshold)"/>
        <param name="iou_threshold" value="$(arg iou_threshold)" />
        <param name="maximum_detections" value="$(arg maximum_detections)"/>
        <param name="input_image_topic" value="$(arg input_image_topic)"/>
        <param name="view_image" value="$(arg view_image)"/>
        <param name="publish_image" value="$(arg publish_image)"/>
        <param name="output_image_topic" value="$(arg output_image_topic)"/>
        <param name="output_mask_topic" value="$(arg output_mask_topic)"/>
    </node>

    <node pkg="qm_vision" name="masked_depth" type="masked_depth.py" output="screen">
        <param name="view_image" value="$(arg view_image)"/>
        <param name="output_mask_topic" value="$(arg output_mask_topic)"/>
        <param name="input_depth_topic" value="$(arg input_depth_topic)"/>
        <param name="output_masked_depth_nr_topic" value="$(arg output_masked_depth_nr_topic)"/>
    </node>

    <node pkg="nodelet" type="nodelet" name="nodelet_manager" args="manager" />

    <node pkg="nodelet" type="nodelet" name="nodelet1"
        args="load depth_image_proc/point_cloud_xyz nodelet_manager">
        <remap from="camera_info" to="$(arg camera_info_topic)"/>
        <remap from="image_rect" to="$(arg output_masked_depth_nr_topic)"/>
        <remap from="points" to="$(arg output_pc_topic)"/>
    </node>

    <node pkg="qm_vision" name="handle_centroid" type="handle_centroid.py" output="screen">
        <param name="output_pc_topic" value="$(arg output_pc_topic)"/>
        <param name="output_centroid_topic" value="$(arg output_centroid_topic)"/>
        <param name="output_orientation_topic" value="$(arg output_orientation_topic)"/>
    </node>

</launch>